assets/src/edit-story/migration/migrations/v0013_videoIdToId.js

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+function videoIdToId({ pages, ...rest }) {
+  return {
+    pages: pages.map(reducePage),
+    ...rest,
+  };
+}
+
+function reducePage({ elements, ...rest }) {
+  return {
+    elements: elements.map(updateElement),
+    ...rest,
+  };
+}
+
+function updateElement(element) {
+  if (element.resource && 'videoId' in element.resource) {
```

Comment:
This is important part. But just like with reducers, this system requires always to clone things if they change. E.g. this should be:

```
if (doesNotNeedToChange) {
  return element;
}
return {
  ...element,
  resource: {
    ...element.resource,
    id: element.resource.videoId,
    videoId: undefined,
  },
};
```

Note that undefines instead of deletes are totally ok in most of cases here b/c they are auto-deleted on serialization.

Reply:
I was thinking about it.
If this was in the reducer - no doubt. Because we want to re-render if you ever change something in the e.g. resource.

But this is in a simple `migrate` function that needs to preprocess the data, then we restore the story from that data (and this will be the baseline for change detection, no need for re-renders before that).

It's like useState initial value:
```jsx
const storyDataRaw = {
  version: 1,
  pages: [{ elements: [{ resource: { count: 1, toDelete: true } }] }]
};
const migrate = data => {
  // It's ok, less code than cloning
  data.pages[0].elements[0].resource.count = 2;
  data.pages[0].elements[0].resource.toDelete = undefined;
  return data;
};

...
const [storyData] = useState(
  storyDataRaw && migrate(storyDataRaw)
);
```
Is that correct thinking?
I can still follow reducer-style, just want to know is there anything wrong with it?

Reply:
Is this really even needed? We are not even in alpha yet and all this MIGHT break is if the existing video elements on dev boxes right?

Reply:
Correct, we can only break our (dev+staging) meticulously made stories.

For me, it was pre-production training 😉

Reply:
Yeah, I took a look and not seeing anywhere `preMigration === postMigration` checks. So your path is probably ok. Though I still feel weird about reusing objects and the level of effort is very similar.